File: OOPRecords.Model/Teacher.cs
using NakedObjects;
using System.Collections.Generic;

namespace OOPRecords.Model
{
    public class Teacher
    {
        [Hidden(WhenTo.Always)]
        public virtual int Id { get; set; }

        [MemberOrder(1)]
        public virtual string FirstName { get; set; }

        [MemberOrder(2)]
        public virtual string LastName { get; set; }

        [MemberOrder(3)]
        public virtual string JobTitle { get; set; }

        [MemberOrder(4)]
        public virtual ICollection<Student> Tutees { get; set; } = new List<Student>();

        public override string ToString()
        {
            return $"{FirstName} {LastName}, {JobTitle}";
        }
    }
}

File: OOPRecords.ModelVB/Teacher.vb
Imports NakedObjects

Public Class Teacher
    Private _container As IDomainObjectContainer
    Public WriteOnly Property Container() As IDomainObjectContainer
        Set(ByVal value As IDomainObjectContainer)
            _container = value
        End Set
    End Property

    <Hidden(WhenTo.Always)>
    Public Overridable Property Id As Integer

    <MemberOrder(1)>
    Public Overridable Property Title As String

    <MemberOrder(2)>
    Public Overridable Property LastName As String

    <MemberOrder(3)>
    Public Overridable Property JobTitle As String

    <MemberOrder(4)>
    Public Overridable Property Tutees() As ICollection(Of Student)

    <MemberOrder(5)> <Eagerly(EagerlyAttribute.[Do].Rendering)>
    <TableView(False, "Subject", "YearGroup", "SetName")>
    Public Overridable Function SetsTaught() As ICollection(Of TeachingSet)
        Dim id As Integer = Me.Id
        Return _container.Instances(Of TeachingSet)().Where(Function(s) s.Teacher.Id = id).OrderBy(Function(s) s.Subject.Name).ThenBy(Function(s) s.YearGroup).ToList()
    End Function


    Public Overrides Function ToString() As String
        Return $"{Title} {LastName}, {JobTitle}"
    End Function
End Class
